docs/content/sips/000-app-config.md

```diff
@@ -0,0 +1,143 @@
+
+```toml
+[config]
+# This simple form...
+key1 = "default_value"
+# ...is equivalent to:
+key1 = { default = "default_value" }
+# required & secret slot 
+key2 = { required = true, secret = true }
+```
+
+Default values can use template strings to reference other slots.
+```toml
+[config]
+key1 = { required = true }
+key2 = "prefix-{{ key1 }}-suffix"
+```
+
+### Components and applications can set configuration of their direct dependencies
+
+In dependency configuration, templates can reference top-level config keys (those in `[config]`), "sibling" keys within the same dependency, and "ancestor" dependant configs.
+
+- Top-level references use just the key name: `{{ top_level_key }}`
+- "Sibling" references use a single `.` prefix: `{{ .sibling_key }}`
+- "Ancestor" references use multiple `.`s: `{{ ..parent_dep_key }}`, `{{ ...grandparent_key }}`
+
+`spin.toml`:
+```toml
+[config]
+app_root = "/app"
+log_file = "{{ app_root }}/log.txt"    # -> "/app/log.txt"
+...
+[[component.config]]
+work_root = "{{ app_root }}/work"      # -> "/app/work"
+work_out = "{{ .work_root }}/output"   # -> "/app/work/output"
+[[component.dependencies.dep1.config]]
+dep_root = "{{ ..work_root }}/dep"     # -> "/app/work/dep"
+```
+
+### Configuration "providers" resolve application configuration
+
+When resolving the value of an application configuration slot, providers will be queried in-order for a value. If no value is returned by any provider, the resolution will either use the default value or fail (if the slot is "required").
+
+Provider configuration is handled by spin at instantiation time (`spin up`).
+
+_Note: Provider configuration is TBD; as TOML it could look like:_
+
+```toml
+[[config-provider]]
+type = "json_file"
+path = "config.json"
+
+[[config-provider]]
+type = "env"
+prefix = "MY_APP_"
+```
+
+#### Example providers
+
+- Environment provider
+  - Configured with a prefix, e.g. `SPIN_APP_`
+  - `key_one` -> `SPIN_APP_KEY_ONE`
+- File provider
+  ```json
+  {"key_one": "value-one"}
+  ```
+- Vault provider
+
+### Configuration is exposed via component interface
+
+`spin-config.wit`
+```
+// Unknown key is a runtime error
+get-config: function(key: string) -> expect<string>
+```
+- Since each component gets its own instance of the `spin-config` import, the executor can resolve keys automatically and only expose a component's own config to it.
+
+### Future design options
+
+#### Typed configuration
+
+The above assumes only string values, but we could include some typing:
+```toml
+# Simple form is typed implicitly from its default value
+string_key = "value"
+string_key = { type = "string", default = "value" }
+
+number_key = 123
+number_key = { type = "number", default = 123 }
+
+required_string = { type = "string", required = true }
+# "bytes" would require e.g. base64 encoding in some places
+encryption_key = { type = "bytes", required = true, secret = true}
+```
+- This would complicate the implementation but might be nice for users.
```

Comment:
Expected types and conversion semantics might vary from language to language too, not sure.  Like "number" - what does that translate to in JS vs AssemblyScript vs C?

Reply:
Everything down here is under the "Future design options" section, so I'm not trying to fully specify it. My vague idea on types is that they would be :wave:mostly JSON-compatible:wave: (perhaps excluding `null` and including some non-unicode byte string type), thus punting the problem to JSON libraries.